### DISCUSSION TITLE
Is there a good CRUD example application?

### DISCUSSION THREAD

**Discussion by jannejava** [livewire/livewire#3478] [1. Help]
HI! 

I find this project very interesting. I am starting to learn and evaluate Livewire. One thing I struggle with is a demo application on how to structure a more complex CRUD Livewire app. I am not talking about counters and to-do lists since I get the concepts but want to move to the next level.

For example, a simple thing: how do you structure a CRUD app with an index page and a detail view? The back button gets broken in Safari if you use `redirect()` to view the details, and you have to mess with the caching. I solved this with modals for now, but how do others do this?

Currently, I am also looking at inertiajs.com, and there is a demo app like: https://inertiajs.com/demo-application

If you have a repo or other tips, I would very much appreciate it.

/ Jan

**Comment by joshhanley** [comment]
@jannejava not public, but if you sponsor Caleb, you get access to the screen casts and the "Surge" repo which he builds in the screencasts, that covers all of that.

In regards to back button stuff, we've been working on a solution that will hopefully solve that!

Hope this helps!

**Reply by jannejava** [reply]
Thanks for explaining, Josh! I appreciate it!

I thought the git sponsor video was only for advanced data tables and so forth, but it makes sense. I didn't think of sponsoring since I was dipping my toes in Limewire, but that seems the way to go then.

The lack of actual other 3rd party examples and tutorials on the web is lacking. It seems most of them are just copied from the same tutorial for SEO-crap – plain bad or outdated.

The back button issue in Safari is a bit of a show stopper for me. It sounds great that you are addressing it!

**Reply by joshhanley** [reply]
@jannejava no worries 🙂

Yeah it's a pretty comprehensive video series with more than just datatables, so definitely worth it.

Agreed, one disadvantage to Livewire being so young!

**Reply by joowdx** [reply]
Hi @joshhanley! 

If you don't mind me asking, how do we force livewire to refresh the page when back button is clicked? I've read somewhat similar, if not the same issue, but I can't happen to find it again so I'm asking here 😅

Thanks!

**Reply by jannejava** [reply]
Shouldn’t that be automatic (except for Safari where you need to some clever workarounds)?

**Reply by joowdx** [reply]
Not really sure. I'm nesting components and using queryStrings . Also I'm using a full page component and I think if any time queryStrings are changed, you get to click back at that previous specific state. 

My only problem is that it doesn't update anything outside the component, using slots, back to its previous state.

Sample component view.
```
<div>
    {{-- Changes on this named slot don't get updated when clicking back button --}}
    <x-slot name="header">
        Header
    </x-slot/>
    ....

    {{-- Trigger changes in header --}}
    <button x-data={} @click="$wire.emit('update-header', 'new header text', 'next header text after clicking back')">
        Change header
    </button>
</div>
```

Header component view
```
<header id="header" class="sticky top-0 z-10 text-xl font-semibold leading-tight text-gray-800 bg-white shadow dark:bg-gray-800 dark:text-gray-100">
    <div class="flex justify-between px-4 py-3 mx-auto max-w-7xl sm:px-6 lg:px-8">
        <h1 class="py-1">
            {!! $header ?? '&nbsp;' !!}
        </h1>
        @if ($back)
            <x-jet-button class="flex-none py-1" wire:click="back">
                Go back
            </x-jet-button>
        @endif
    </div>
</header>

```

Header controller
```
<?php

namespace App\Http\Controllers;

use Livewire\Component;

class Header extends Component
{
    public $header, $back;

    protected $listeners = [
        'update-header' => 'updateHeader',
    ];

    public function mount($header, $back)
    {
        $this->header = $header ? $header->toHtml() : null;
        $this->back = $back ? $back->toHtml() : null;
    }

    public function render()
    {
        return view('header');
    }

    public function updateHeader($header, $back = null)
    {
        $this->header = $header;
        $this->back = $back;
    }

    public function back()
    {
        $this->emit('back');
        $this->header = $this->back;
        $this->back = null;
    }
}

```

I don't honestly know what's the best approach for this. Just making my own and it works fine. My only problem is that when clicking the back button. I know there's an issue about the *browser* back button, what I don't know if this is related, just wanna know how can we do some workarounds 😅

**Reply by joshhanley** [reply]
@joowdx currently both chrome and safari cache browser state (but in different ways), so each display stale state when you click back. There is some hacks you can find one a couple of issues (before we moved to discussions), which is if you search for back button or cache or something should come up.

There is a permanent solution in the works, which should be released in the next version. You can check out the PR #2747.

@jannejava yeah it should be automatic, but it's not, yet 😅 and it's more than just safari.

But we think we have managed to solve it for the different browsers and each of their different issues.

Hope this helps!

**Reply by joowdx** [reply]
@joshhanley uhm just another question... by chrome does that also refer to any chromium based browsers?

**Reply by joshhanley** [reply]
@joowdx yep, I believe so (although I haven't tested extensively)

**Comment by joowdx** [comment]
Thanks a bunch @joshhanley ! You rock \m/

**Reply by joshhanley** [reply]
Thanks! 😊
